### ISSUE TITLE
Investigate PropTypes support

### ISSUE THREAD

**Issue by mikekasprzak** [JammerCore/JammerCore#1568]
https://reactjs.org/docs/typechecking-with-proptypes.html

This is a react thing, but are interesting elements to it.

* type checking for prop values passed to components
* setting `defaultProps` via a simplified syntax

In our case we'll have to skip out on `defaultProps`, but the idea of a defaults overlay is something we could look in to implementing in a different way.

Type checking is interesting because... well, it'll help us find bugs.

The added code could be wrapped in an ifdef block.

```js
// @ifdef DEBUG
only_do_this_in_debug_build();
// @endif
```

This way there's no overhead added to the runtime app, and we get all the benefits of type checking.

**Comment by mikekasprzak** [comment]
NB: do the import at the bottom of the file, alongside the type checking. This way you only need a single `@ifdef` block.

**Comment by local-minimum** [comment]
Common complex types that we often shuffle between components should be defined in a separate file, e.g. `prop-types.js` and there we would have things like

```
const nodeType = PropTypes.shape({
   slug: PropTypes.string.isRequired,
   ...
});
```

**Comment by mikekasprzak** [comment]
This is indirectly resolved by #1998 introducing relaxed Typescript checking within VSCode. That said that exploration continues here #2065.
